File: ios/Runner/LogBridge.swift
import Flutter
import UIKit

#if DEBUG

/// DEBUG-only log bridge that captures stdout/stderr and forwards them
/// over a Flutter MethodChannel so they can be written to the native-ios
/// log file alongside rust and dart logs.
final class LogBridge {
    private let channel: FlutterMethodChannel
    private var pipeReadFd: Int32 = -1
    private var backgroundThread: Thread?

    init(messenger: FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(
            name: "heart_beat/native_log",
            binaryMessenger: messenger
        )
        channel.setMethodCallHandler { [weak self] call, result in
            // Handlers can be added here if Dart needs to query bridge state.
            result(FlutterMethodNotImplemented)
        }

        startCapture()
    }

    private func startCapture() {
        // Create a pipe for stdout and stderr.
        var pipefd = [Int32](repeating: 0, count: 2)
        guard pipe(pipefd) == 0 else {
            print("[LogBridge] pipe() failed: \(errno)")
            return
        }

        let readFd = pipefd[0]
        let writeFd = pipefd[1]

        // Dup the write end to both stdout and stderr.
        guard dup2(writeFd, STDOUT_FILENO) != -1,
              dup2(writeFd, STDERR_FILENO) != -1 else {
            print("[LogBridge] dup2() failed: \(errno)")
            close(pipefd[0])
            close(pipefd[1])
            return
        }

        // Close the original writeFd — we've already duplicated it.
        close(writeFd)

        pipeReadFd = readFd

        // Spawn a background thread to drain the pipe.
        backgroundThread = Thread { [weak self] in
            self?.readLoop(readFd: readFd)
        }
        backgroundThread?.name = "LogBridge.readLoop"
        backgroundThread?.qualityOfService = .utility
        backgroundThread?.start()
    }

    private func readLoop(readFd: Int32) {
        let bufferSize = 4096
        var buffer = [UInt8](repeating: 0, count: bufferSize)

        while true {
            let n = read(readFd, &buffer, bufferSize)
            if n <= 0 {
                break
            }

            // Convert to String, split on newlines, send each line.
            let data = Data(bytes: buffer, count: n)
            if let text = String(data: data, encoding: .utf8) {
                let lines = text.components(separatedBy: "\n")
                for line in lines where !line.isEmpty {
                    DispatchQueue.main.async { [weak self] in
                        self?.channel.invokeMethod("onNativeLog", arguments: line)
                    }
                }
            }
        }

        close(readFd)
    }

    func dispose() {
        // Closing the read end will cause readLoop to exit.
        if pipeReadFd != -1 {
            close(pipeReadFd)
            pipeReadFd = -1
        }
        backgroundThread?.cancel()
        backgroundThread = nil
    }

    deinit {
        dispose()
    }
}

#endif

File: android/app/src/main/kotlin/com/example/heart_beat/LogBridge.kt
package com.example.heart_beat

import android.os.Build
import io.flutter.embedding.android.FlutterActivity
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel

#if DEBUG
/// DEBUG-only log bridge that captures logcat output and forwards it
/// over a Flutter MethodChannel so lines can be written to the native-android
/// log file alongside rust, dart, and native-ios logs.
class LogBridge(private val activity: FlutterActivity) {
    private val channel = MethodChannel(
        activity.binaryMessenger,
        "heart_beat/native_log"
    )

    private var logcatProcess: Process? = null
    private var backgroundThread: Thread? = null

    fun start() {
        try {
            // Get the current process ID
            val pid = android.os.Process.myPid()

            // Start logcat with threadtime format (includes timestamp, thread, tag)
            // Filter to only this app's PID
            logcatProcess = Runtime.getRuntime().exec(
                arrayOf("logcat", "-v", "threadtime", "--pid=$pid")
            )

            backgroundThread = Thread {
                readLoop()
            }
            backgroundThread?.name = "LogBridge.readLoop"
            backgroundThread?.priority = Thread.MIN_PRIORITY
            backgroundThread?.start()
        } catch (e: Exception) {
            android.util.Log.e("LogBridge", "Failed to start logcat: $e")
        }
    }

    private fun readLoop() {
        try {
            val reader = logcatProcess?.inputStream?.bufferedReader() ?: return

            var line: String?
            while (true) {
                line = reader.readLine()
                if (line == null) break

                // Skip empty lines
                if (line.isEmpty()) continue

                // Forward to Flutter via main thread
                activity.runOnUiThread {
                    try {
                        channel.invokeMethod("onNativeLog", line)
                    } catch (e: Exception) {
                        // Channel might not be ready yet, ignore
                    }
                }
            }
        } catch (e: Exception) {
            android.util.Log.e("LogBridge", "readLoop error: $e")
        } finally {
            try {
                logcatProcess?.destroy()
            } catch (_: Exception) {
            }
        }
    }

    fun dispose() {
        backgroundThread?.interrupt()
        backgroundThread = null

        try {
            logcatProcess?.destroy()
            logcatProcess = null
        } catch (_: Exception) {
        }
    }
}
#endif